raft topology: ban left nodes from the cluster

kbr-scylla opened the pull request:
Use the new Seastar functionality for storing references to connections to implement banning hosts that have left the cluster (either decommissioned or using removenode) in raft-topology mode. Any attempts at communication from those nodes will be rejected.

This works not only for nodes that restart, but also for nodes that were running behind a network partition and we removed them. Even when the partition resolves, the existing nodes will effectively put a firewall from that node.

Some changes to the decommission algorithm had to be introduced for it to work with node banning. As a side effect a pre-existing problem with decommission was fixed. Read the "introduce `left_token_ring` state" and "prepare decommission path for node banning" commits for details.

kbr-scylla commented:
In the meantime, I'll working on some test infrastructure to test this.
(I already checked manually that it works)

gleb-cloudius commented on message/messaging_service.cc:
```diff
@@ -284,6 +284,35 @@ bool messaging_service::is_same_rack(inet_address addr) const {
+        }
+
+        auto [start, end] = ms._host_connections.equal_range(id);
+        std::vector<shared_ptr<rpc::server::connection>> conns;
```
I do not like that we expose an internal connection object to an application like that. OTOH we already did it in `rpc::server::foreach_connection` API, but at least it only gives a const access.

kbr-scylla replied:
How about `rpc::server::drop_connection(connection_id)`?
We'll still have to store the IDs and `rpc::server&`, but we won't need to work with `server::connection`.

Doesn't change much from the constness side though, we still allow "write access" in this way, just in smaller scope.

gleb-cloudius replied:
Smaller scope it better. I also thought about `drop_connection`. Does it have to be synchronous BTW or can it be void?

kbr-scylla replied:
I think it can be void.

kbr-scylla replied:
Hm, I still need to obtain a reference to non-const `server` (since `drop_connection` will be non-const) here:
```cpp
    register_handler(this, messaging_verb::CLIENT_ID, [this] (rpc::client_info& ci, gms::inet_address broadcast_address, uint32_t src_cpu_id, rpc::optional<uint64_t> max_result_size, rpc::optional<utils::UUID> host_id) {
        if (host_id) {
            auto peer_host_id = locator::host_id(*host_id);
            ci.attach_auxiliary("host_id", peer_host_id);
            _host_connections.emplace(peer_host_id, connection_ref {
                .server = ci.conn.get_server(),
                .conn_id = ci.conn.get_connection_id(),
```

So it looks like we need the `connection&` inside `client_info`.

kbr-scylla replied:
Alternatively, I could store `rpc::server&` + `connection_id` inside `client_info`, which is two fields instead of one `connection&`, but maybe that would be better? Dunno.

gleb-cloudius replied:
You already store both of them in ` messaging_service::connection_ref` anyway.

kbr-scylla commented on message/messaging_service.cc:
```diff
@@ -284,6 +284,35 @@ bool messaging_service::is_same_rack(inet_address addr) const {
+        ms._host_connections.erase(start, end);
+
+        co_await parallel_for_each(conns, [] (shared_ptr<rpc::server_connection>& conn) {
+            return conn->stop();
```
We could, but if some service is sending out RPCs to that node, `get_rpc_client` would just reestablish the connection.

I guess we could also ban outgoing connections like we do with incoming (and checking whether a host is banned in `get_rpc_client` or something).

Alternatively - and this is the approach I took basically - assume that once a node has left, *we* will not attempt to communicate with it (fingers crossed). This depends on nice behavior of our services. So we don't have to do the firewall in this direction, because we're not sending anything to that node anymore anyway.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1184/

xemul commented on main.cc:
```diff
@@ -1248,6 +1248,10 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
+            messaging.invoke_on_all([host_id = cfg->host_id] (netw::messaging_service& ms) {
+                ms.set_my_host_id(host_id);
+            }).get();
+
```
Is it possible to move this into `sys_ks.local().setup(snitch, messaging).get();` below the main code? All quirks to straightforward system-keyspace initialization are currently there and I'd like not to spread them around.

kbr-scylla replied:
The only important thing for `set_my_host_id` is that we call it before we start sending any message.

kostja replied:
Since host id is optional, we need an assert in establish-connection path that this key is set before the service is used to open outgoing connections.

kbr-scylla commented:
> General question -- is seastar part really needed? There's messaging_service::find_and_remove_client (and stuff on top) that can be used to drop the connection without the help of seastar. Would it work here?

IIUC it only works for clients, i.e. outgoing connections/messages. The seastar part is needed for blocking incoming messages.

+ dropping existing connections is not enough, we also need a way to prevent further communication.

kostja commented on message/messaging_service.cc:
```diff
@@ -276,6 +284,35 @@ bool messaging_service::is_same_rack(inet_address addr) const {
+    });
+}
+
+bool messaging_service::host_banned(locator::host_id id) {
```
Nit: is_host_banned would look nicer for a predicate.

kostja commented on message/messaging_service.cc:
```diff
@@ -384,7 +421,18 @@ messaging_service::messaging_service(config cfg, scheduling_config scfg, std::sh
+        if (host_id) {
+            auto peer_host_id = locator::host_id(*host_id);
+            if (host_banned(peer_host_id)) {
+                throw rpc::drop_connection{};
```
Can't you just close the connection here right away, why do you need to propagate the failure via an exception?

kbr-scylla commented:
> do we really need to be able to stop the connection from within reply() if rpc::drop_connection is thrown? Do we need drop-on-exception, isn't explicit drop enough?

Without the exception it would be awkward if a two-way RPC handler wants to drop a connection. The handler would stop the connection but then it would still have to construct some kind of response or throw another type of exception.

For this use case though we only close connections from one-way handlers so maybe the exception is not necessary.

> stop() calls abort(), which calls shutdown_input() which only calls shutdown (SHUT_RD). Why only read part? Why not close the file descriptor altogether?

That's a question for @gleb-cloudius

gleb-cloudius commented:
> > stop() calls abort(), which calls shutdown_input() which only calls shutdown (SHUT_RD). Why only read part? Why not close the file descriptor altogether?
> 
> That's a question for @gleb-cloudius

Shutdown read wake ups a reader fiber which start shutdown process, notifies sender fiber which shutdowns write side at the end and closes the descriptor.

kbr-scylla commented:
v2:

seastar:
- in `rpc::client_info&`, provide `rpc::server&` + `connection_id` instead of `server::connection&`
- replace `server::get_connection` with `server::drop_connection(connection_id)`
- remove `rpc::drop_connection` exception; the handler has access to the connection from other commits in this series, it can drop it directly
- remove "rpc: move `server::connection` class outside `server`" commit - not needed with the new approach
- add a simple test for dropping connection from handler

scylla:
- use `client_info::server` and `client_info::conn_id` fields (instead of `client_info::conn` which was removed from the seastar patch)
- use `server::drop_connection(conn_id)` instead of `connection::drop()` (we don't have access to the `connection` anymore)
- instead of throwing `rpc::drop_connection{}` exception in handler, use `server::drop_connection` - the exception API was redundant
- rename `host_banned` to `is_host_banned`
- add test infrastructure + simple test for banning nodes

I didn't implement dropping/banning of outgoing connections, because it's blocked by #6403. When we send a message, in general, we don't have access to the `host_id` of the destination inside `messaging_service`.
AFAIK @bhalevy is working on extending `msg_addr` with `host_id`s. Then it will be possible.

Still, I think that banning incoming connections provides 99%-100% of the value that this PR is supposed to provide: it effectively isolates removed nodes, as the added test illustrates.

I propose that in the scope of this PR, we solve banning incoming connections, and leave banning outgoing connections to a follow-up, after we have #6403.

Unfortunately, there's a bigger problem.
Decommissioning doesn't work!

Well, it does to some extent, the node is removed from token ring / topology... but it cannot remove itself from group 0, because by that time, it is banned from the cluster :D cc @kostja @gleb-cloudius
This is illustrated by `test_topology_ops` which is now failing.

We should discuss how to solve this. This IMO supports the theory that the topology coordinator should be responsible for group 0 reconfigurations.

gleb-cloudius commented:
> but it cannot remove itself from group 0, because by that time, it is banned from the cluster

But the coordinator already does that in case a node itself fails. So may be we only need to silence the error?

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1253/

kbr-scylla commented:
> But the coordinator already does that in case a node itself fails. So may be we only need to silence the error?

The "error" is that the decommissioning node hangs on `leave_group0()`.

kbr-scylla commented:
> Then we need to do it with a timeout.

What's the point of having a call which will timeout 90% of the time? We expect it to fail, so why do it in the first place?

gleb-cloudius commented:
> > Then we need to do it with a timeout.
> 
> What's the point of having a call which will timeout 90% of the time? We expect it to fail, so why do it in the first place?

True. We can not do it at all.

kbr-scylla commented:
I don't like the resulting UX.

`nodetool decommission` should give certain guarantees to the user when it finishes successfully. It should not regress. Currently we are guaranteed that once it finishes, the node is no longer part of topology or group 0 configuration. We should not break this promise.

So IMO the decommissioning node should get a confirmation from the topology coordinator that the node got removed entirely.

But when we ban left nodes, we make it impossible for the node to get this confirmation. Once it's removed, we're no longer communicating with it, but we still want to tell it that it's removed.

kbr-scylla commented:
> When nodetool decommission completes there is no guaranty the node in not part of group0 and that is why you added a clean up procedure to nodetool removenode.

The guarantee is there if it completes successfully.
The cleanup procedure is only needed if decommission fails.

> The only other way to "fix" ux is to issue nodetool decommission to another node and add a parameter to provide node id to decommission. IOW make it like nodetool removenode for live nodes.

Or ban the node only after it gets the confirmation.

gleb-cloudius commented:
> > When nodetool decommission completes there is no guaranty the node in not part of group0 and that is why you added a clean up procedure to nodetool removenode.
> 
> The guarantee is there if it completes successfully. The cleanup procedure is only needed if decommission fails.

With the topology coordinator cleanup procedure is not needed period. So where is a regression?

> 
> > The only other way to "fix" ux is to issue nodetool decommission to another node and add a parameter to provide node id to decommission. IOW make it like nodetool removenode for live nodes.
> 
> Or ban the node only after it gets the confirmation.

How can you know it got it?

kbr-scylla commented:
> With the topology coordinator cleanup procedure is not needed period. So where is a regression?

Today (without experimental flag etc.), if decommission completes successfully, the user knows that the node was already removed from group 0 configuration and token ring.

If we no longer guarantee that, it will be a regression.

> How can you know it got it?

Yeah I don't see how.

Maybe we shouldn't ban decommissioning nodes? After all they are leaving gracefully, they know they are leaving and won't try rejoin the cluster. Banning makes sense for `removenode` where we can remove a node that doesn't know about it.

gleb-cloudius commented:
> > With the topology coordinator cleanup procedure is not needed period. So where is a regression?
> 
> Today (without experimental flag etc.), if decommission completes successfully, the user knows that the node was already removed from group 0 configuration and token ring.

The user does not know what group0 is. What practical different to you see?

> 
> If we no longer guarantee that, it will be a regression.
> 
> > How can you know it got it?
> 
> Yeah I don't see how.
> 
> Maybe we shouldn't ban decommissioning nodes? After all they are leaving gracefully, they know they are leaving and won't try rejoin the cluster. Banning makes sense for `removenode` where we can remove a node that doesn't know about it.

Why wouldn't they won't try? We cannot guaranty that.

kbr-scylla commented:
> The user does not know what group0 is. What practical different to you see?

One thing is that it makes testing harder. Now I can't do the immediate check "node is not a member of group 0" after decommission finishes.

But even if we don't care about group 0 config, with banning we cannot even tell the node that it left token ring / topology.
I guess this could be handled by having a last step in the decommission saga which we cannot abort - if we're in this step we can only finish the decommission - and before banning the node, we wait for it to learn that we're in this step (e.g. using read barrier), which causes it to finish and shutdown, then we finalize removal and ban it.

> Why wouldn't they try? We cannot guaranty that.

By persisting a flag that the node is decommissioned so it will refuse to reboot (we already do that).

gleb-cloudius commented:
> > The user does not know what group0 is. What practical different to you see?
> 
> One thing is that it makes testing harder. Now I can't do the immediate check "node is not a member of group 0" after decommission finishes.

We need to test what happens if node dies before removing itself from config anyway.

> 
> But even if we don't care about group 0 config, with banning we cannot even tell the node that it left token ring / topology. I guess this could be handled by having a last step in the decommission saga which we cannot abort - if we're in this step we can only finish the decommission - and before banning the node, we wait for it to learn that we're in this step (e.g. using read barrier), which causes it to finish and shutdown, then we finalize removal and ban it.
> 
> > Why wouldn't they try? We cannot guaranty that.
> 
> By persisting a flag that the node is decommissioned so it will refuse to reboot (we already do that).

OK. But someone may claim that removing all or some date and restarting the node suppose to work because there is a dtest that did that and worked by miracle. But on a more serious note I prefer uniformity: if a node is in the `left` state for any reason it is banned.

kostja commented:
We need the banned node to acknowledge banning, then ban it (success path).
The failure path (no last acknoledgment) should ban after a timeout.

gleb-cloudius commented:
> > 
> > By persisting a flag that the node is decommissioned so it will refuse to reboot (we already do that).
> 
> OK. But someone may claim that removing all or some date and restarting the node suppose to work because there is a dtest that did that and worked by miracle. But on a more serious note I prefer uniformity: if a node is in the `left` state for any reason it is banned.

Found this gem in our code:

```
    if (_sys_ks.local().was_decommissioned()) {
        if (_db.local().get_config().override_decommission()) {
            slogger.warn("This node was decommissioned, but overriding by operator request.");
            co_await _sys_ks.local().set_bootstrap_state(db::system_keyspace::bootstrap_state::COMPLETED);
        } else {
            auto msg = sstring("This node was decommissioned and will not rejoin the ring unless override_decommission=true has been set,"
                               "or all existing data is removed and the node is bootstrapped again");
            slogger.error("{}", msg);
            throw std::runtime_error(msg);
        }
    }

```

So today  decommissioned node can be in fact restarted. The question is do we want to support that with raft? Even today with schema over raft only this option is broken and at least needs to be fixed to fail if raft is enabled.

bhalevy commented:
> So today decommissioned node can be in fact restarted. The question is do we want to support that with raft? Even today with schema over raft only this option is broken and at least needs to be fixed to fail if raft is enabled.

I've never seen this feature being used. I'm not even aware of a test for it, so I think there's no guarantee it even works.
I think that after decommissioned is "committed" there is no point in bringing the node back to life since it already lost its tokens to other nodes.  If the node is fixed and we want to get its tokens back, we should reinitialize it so to wipe its data and erase its previous host_id (by erasing the system tables) and then bootstrap it.
It will get different tokens in this case, but I don't think it matters much.  Is there any benefit in restoring the tokens it had before being decommissioned?

cc @avikivity @tgrabiec

mykaul commented:
Isn't https://docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/revoke-decommission.html the procedure for it?

kbr-scylla commented:
> Isn't https://docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/revoke-decommission.html the procedure for it?

This effectively bootstraps a completely fresh node, because data and commitlog dirs are removed:
```
sudo rm -rf /var/lib/scylla/data
sudo find /var/lib/scylla/commitlog -type f -delete
sudo find /var/lib/scylla/hints -type f -delete
sudo find /var/lib/scylla/view_hints -type f -delete
```

The `override_decommission` option that Gleb found makes the node simply restart with its old tokens without doing the regular bootstrap procedure (which involves streaming/repair during token movement etc.) which will just cause data loss.

It's another one of those options which we inherited from Cassandra and were (hopefully) never used in ScyllaDB production clusters.

kbr-scylla commented:
v3:
- some changes to the decommission algorithm had to be introduced for it to work with node banning. As a side effect a pre-existing problem with decommission was fixed. Read the "introduce `left_token_ring` state" and "prepare decommission path for node banning" commits for details.

All tests should now pass.

I also opened a PR in Seastar: https://github.com/scylladb/seastar/pull/1663

kbr-scylla commented:
Removed RFC: prefix, but we'll have to wait for https://github.com/scylladb/seastar/pull/1663 to get merged before rebasing and then merging this PR.

kbr-scylla commented:
@benipeled do you know what's happening to the build?
It looks like it hanged: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1376/console
last output is
```
00:26:32      "received_events_url": "https://api.github.com/users/scylladb-promoter/received_events",
00:26:32      "type": "User",
00:26:32      "site_admin": false
00:26:32    }
00:26:32  }
00:26:32  [Pipeline] }
00:26:32  [Pipeline] // withCredentials
00:26:32  [Pipeline] }
00:26:32  [Pipeline] // script
00:26:32  [Pipeline] }
00:26:32  [Pipeline] // stage
00:26:32  [Pipeline] }
```

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1376/

kbr-scylla commented:
It looks like the dtest gating job never managed to get scheduled
https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/1624/console
```
23:55:29  Still waiting to schedule task
23:55:29  All nodes of label ‘[ec2-fleet-4cpu-dtest-asg-spot](https://jenkins.scylladb.com/label/ec2-fleet-4cpu-dtest-asg-spot/)’ are offline
09:30:53  Calling Pipeline was cancelled
09:30:53  [Pipeline] // node
```

gleb-cloudius commented on db/system_keyspace.cc:
```diff
@@ -3657,9 +3657,14 @@ future<service::topology> system_keyspace::load_topology_state() {
-            on_internal_error(slogger,
-                "load_topology_state: topology not in transition state but transition nodes are present");
+        } else {
+            for (auto& [id, rs]: ret.transition_nodes) {
```
Can we use `std::all_of` here?

kbr-scylla replied:
I want to print the ID of the node that's in the wrong state.
I changed it to `find_if`

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1128,6 +1130,21 @@ class topology_coordinator {
+                // Wait until other nodes observe the new token ring and stop sending writes to this node.
+                bool exec_command_res;
+                std::tie(node, exec_command_res) = co_await exec_global_command(
+                        std::move(node), raft_topology_cmd::command::barrier, false);
```
We need to wait for all writes to complete, no?

kbr-scylla replied:
Yes.

So we need another command like `fence_old_writes`?

gleb-cloudius replied:
We need to rebase on @gusev-p fencing series that already have the drain verb.

kbr-scylla replied:
Or leave a FIXME and merge in a follow-up :P

gleb-cloudius commented on service/topology_state_machine.hh:
```diff
@@ -127,6 +128,8 @@ struct raft_topology_cmd {
       };
       command cmd;
+
+      raft_topology_cmd(command c) : cmd(c) {}
```
Explicit? But then we do not need it.

kbr-scylla replied:
Being implicit is the entire purpose so I can pass e.g. `raft_topology_cmd::command::barrier` instead of `raft_topology_cmd{raft_topology_cmd::command::barrier}` to functions that take `const raft_topology_cmd&`

gleb-cloudius replied:
Such casts may cause surprises. May be not for the rarely used type like here, but we do not save a lot of typing with explicit cast anyway. We also have it in all other places.

kbr-scylla replied:
I really don't like the redundant typing.
How about I extract the constructor to a separate commit and change the code to use it everywhere

gleb-cloudius replied:
One may say it is a redundant typing, but the other may note that this is type safety. I know you like type safety as well. But OK, this is internal type with narrow use. We will manage.

gleb-cloudius commented on message/messaging_service.cc:
```diff
@@ -284,6 +284,32 @@ bool messaging_service::is_same_rack(inet_address addr) const {
     return topo.get_rack(addr) == topo.get_rack();
 }
 
+future<> messaging_service::ban_host(locator::host_id id) {
```
IIRC in the last review I asked to drop client as well. This will be used for node isolation as well.

kbr-scylla replied:
And I answered that I cannot do this in this PR because I'm blocked by https://github.com/scylladb/scylladb/issues/6403
In this comment https://github.com/scylladb/scylladb/pull/13850#issuecomment-1548079128:

> I didn't implement dropping/banning of outgoing connections, because it's blocked by https://github.com/scylladb/scylladb/issues/6403. When we send a message, in general, we don't have access to the host_id of the destination inside messaging_service.
AFAIK @bhalevy is working on extending msg_addr with host_ids. Then it will be possible.

> Still, I think that banning incoming connections provides 99%-100% of the value that this PR is supposed to provide: it effectively isolates removed nodes, as the added test illustrates.

> I propose that in the scope of this PR, we solve banning incoming connections, and leave banning outgoing connections to a follow-up, after we have https://github.com/scylladb/scylladb/issues/6403.

kbr-scylla replied:
Even if I could to this (which I don't see how without #6403), it's IMO better to develop this in a bunch of smaller PRs instead of one giant PR that does everything.

gleb-cloudius replied:
I missed your answer (PRs are crap), but I do not understand why can't we map ids to ips as interim solution?
If we do have a mapping adding the code that drops client is easy. IIRC we already have a function that does it. But OK. Lets leave it for the very near future (we will need it for `ignore-nodes` option.

kbr-scylla replied:
> I missed your answer (PRs are crap), but I do not understand why can't we map ids to ips as interim solution?

It would not work very well in presence of changing IPs or losing IP mappings - for example
1. we ban host A which is mapped to IP X
2. at some point another host B reuses IP X
3. we get message from IP X, our mapping says it's host A which is banned

or, when we remove host A, we eventually stop gossiping its IP and lose its mapping, so we'd also lose the ban

I think such interim solution would bring more problems than it solves issues, we really need to fix #6403, definitely before we GA topology.

gleb-cloudius replied:
#6403 will not magically use UUIDs on the wire.  The scenario you described is exactly what can happen with this patch, but since you do not use ip to check what is banned, but the UUID that you get as part of CLIEN_ID the code works as indented though #6403 is not solved. What is missing is dropping clients as well since you cannot map directly from UUID to a client. The worst that can happen (and I am not sure it can) you will map ID to the wrong IP an drop incorrect client. New one will be created (though some RPCs may fail why they should not have).

kbr-scylla replied:
> The scenario you described is exactly what can happen with this patch, but since you do not use ip to check what is banned, but the UUID that you get as part of CLIEN_ID the code works as indented

So you're saying that the scenario can happen with this patch, but actually it cannot? :)

To drop clients today I would have to map from IPs to UUIDs, not the other way. And this suffers from the problems I described.
The correct solution is to identify clients by UUIDs in the first place. We don't even need UUIDs for the write for that. We just need to know who we're sending the message to.
Which we will be able to do after #6403:
>  AFAIK @bhalevy is working on extending msg_addr with host_ids. Then it will be possible.

gleb-cloudius replied:
> > The scenario you described is exactly what can happen with this patch, but since you do not use ip to check what is banned, but the UUID that you get as part of CLIEN_ID the code works as indented
> 
> So you're saying that the scenario can happen with this patch, but actually it cannot? :)

I am saying it can, but it does not produce the effect you claim because we do not ban by an IP, but by an ID.

> To drop clients today I would have to map from IPs to UUIDs, not the other way.

And the worst that may happen (as I described in my reply) you will drop incorrect client and it will be re-created.
And I am not even sure it may eventually happen because new node will not be able to claim old IP until we complete previous operation which includes banning. May be possible with replace op.

> And this suffers from the problems I described.

No. It does not because we ban by ID, not IP. We only drop clients by IP.

kbr-scylla replied:
So you suggest we only drop clients, but don't ban outgoing messages? What's the value of that, if the client will be recreated automatically on the next message?

gleb-cloudius replied:
Right, for banning outgoing message we need to translate IP to ID on client creation or at least have @bhalevy's patch. But is it a problem? The cluster should be able to learn new mapping through gossiper. We allow incoming connections.

kbr-scylla replied:
IIUC if we get a gossip SYN, we don't update our metadata yet, we need to send back ACK and then receive ACK2 to update the metadata. So we also need to allow the outgoing ACK.

But maybe we could update the mapping earlier - immediately when we receive CLIENT_ID verb.
This would solve the potential availability problem.

But there's also the problem that we're not keeping mappings of nodes that we removed forever - they are TTLd ~1 hour. So after ~1 hour of banning a node, we would unban outgoing messages because we would no longer be able to map the IP to ID and notice that the ID is banned.

gleb-cloudius replied:
And we cannot drop message if mapping is not available since a new node will not be able to communicate with a cluster. It becomes complicated but it does not look like https://github.com/scylladb/scylladb/issues/6403 is planned for the near future.

kbr-scylla replied:
#6403 is blocking raft topology from becoming GA, so we'll be able to do the banning properly before becoming GA as well.

avikivity commented:
I gave it a cursory review (not including the Seastar changes), looks good.

kbr-scylla commented:
v4:
- new version of the Seastar patch, `future<> server::drop_connection(id)` was replaced with `void server::abort_connection(id)`
- use `find_if` in `load_topology_state` sanity check
- separate commit to introduce the `raft_topology_cmd` implicit constructor, use it everywhere
- add FIXME near the `left_token_ring` barrier, it needs to be changed to another command once fencing is merged

Diff:
https://github.com/scylladb/scylladb/compare/1c3488985dfd4996995b21716b777700fa3b33d8..cc0d5e6f3641b09b7673e0bb4ddf70801ab40e61

gleb-cloudius commented on message/messaging_service.cc:
```diff
@@ -401,6 +401,17 @@ msg_addr messaging_service::get_source(const rpc::client_info& cinfo) {
+
+locator::host_id messaging_service::get_my_host_id() {
+    if (!_my_host_id) {
+        on_internal_error(mlogger, "messaging_service::get_my_host_id: host ID missing");
```
If not for internal error the whole function would just be `return _my_host_id.value()` :(

kbr-scylla replied:
But if the value is missing, `value()` would throw a cryptic exception that is super hard to debug. (We don't know where the exception originates from)

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1140,6 +1162,35 @@ class topology_coordinator {
+                bool shutdown_failed = false;
+                try {
+                    node = co_await exec_direct_command(std::move(node), raft_topology_cmd::command::shutdown);
+                } catch (...) {
```
Shouldn't we catch and re-throw `term_changed_error`? OTOH `retake_node` below will do it anyway.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1415/

kbr-scylla commented:
Looks like running these tests locally is not trivial
```
$ ./test.py --mode dev object_store
Found 1 tests.
================================================================================
[N/TOTAL]   SUITE    MODE   RESULT   TEST
------------------------------------------------------------------------------
[1/1]      object_store   dev   [ FAIL ] object_store.run.1
------------------------------------------------------------------------------
The following test(s) have failed: object_store/run
Summary: 1 of the total 1 tests failed

$ cat testlog/dev/object_store.run.1.log 
=== TEST.PY STARTING TEST object_store.run.1 ===
export UBSAN_OPTIONS='halt_on_error=1:abort_on_error=1:suppressions=/home/kbraun/dev/scylladb/ubsan-suppressions.supp'
export ASAN_OPTIONS='disable_coredump=0:abort_on_error=1:detect_stack_use_after_return=1'
test/object_store/run --junit-xml=/home/kbraun/dev/scylladb/testlog/dev/xml/object_store.run.1.xunit.xml
=== TEST.PY TEST object_store.run.1 OUTPUT ===
Scylla is: /home/kbraun/dev/scylladb/build/dev/scylla.
Traceback (most recent call last):
  File "/home/kbraun/dev/scylladb/test/object_store/run", line 26, in <module>
    s3_server_address = os.environ['S3_SERVER_ADDRESS_FOR_TEST']
                        ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen os>", line 679, in __getitem__
KeyError: 'S3_SERVER_ADDRESS_FOR_TEST'

```

gleb-cloudius commented:
On Tue, May 23, 2023 at 06:38:15AM -0700, Kamil Braun wrote:
> Looks like running these tests locally is not trivial
> KeyError: 'S3_SERVER_ADDRESS_FOR_TEST'
> 
What? We did not reimplemented S3 API yet just to run the test locally?!

kbr-scylla commented:
The Seastar PR got merged. I will rebase this PR once Seastar submodule is updated in Scylla.

kbr-scylla commented:
Rebased on `master` (Seastar update is in)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1616/

kbr-scylla commented:
@avikivity @xemul ping for merge

kbr-scylla commented:
@avikivity please review/merge
(between pinging Avi and Pavel, I flipped a coin and the coin decided it's Avi)

kbr-scylla commented:
@avikivity please review/merge

kbr-scylla commented:
BTW. this will improve CI stability
(sporadic failure of test_topology_ops caused by decommissioning node shutdown racing with the topology coordinator removing it from group 0; the PR gets rid of the race)

kbr-scylla commented:
@avikivity please review/merge

kbr-scylla commented:
P I N G U S
M A X I M U S
@avikivity

avikivity commented on db/system_keyspace.cc:
```diff
@@ -1471,6 +1471,11 @@ future<> system_keyspace::setup(sharded<locator::snitch_ptr>& snitch, sharded<ne
+    auto host_id = co_await load_local_host_id();
+    co_await ms.invoke_on_all([host_id] (netw::messaging_service& ms) {
+        ms.set_my_host_id(host_id);
+    });
```
It is really silly that system_keyspace has a build dependency on messaging_service.

avikivity commented on message/messaging_service.cc:
```diff
@@ -284,6 +284,26 @@ bool messaging_service::is_same_rack(inet_address addr) const {
 }
 
+future<> messaging_service::ban_host(locator::host_id id) {
+    return container().invoke_on_all([id] (messaging_service& ms) -> future<> {
```
Is invoke_on_all safe with nested coroutines?

Better to make the outer function a coroutine too and use coroutine::lambda to avoid the lambda coroutine fiasco.

kbr-scylla replied:
Unsafety comes from using captures post-yield, which we don't do here.

But the only reason this function is coroutine is because I was yielding in an earlier version of the PR, I no longer do. We can change it to a non-async function at this point. Will follow-up.

avikivity commented on service/topology_state_machine.hh:
```diff
@@ -34,6 +34,7 @@ enum class node_state: uint16_t {
     replacing,           // the node replaces another dead node in the cluster and it data is being streamed to it
     rebuilding,          // the node is being rebuild and is streaming data from other replicas
     normal,              // the node does not do any streaming and serves the slice of the ring that belongs to it
+    left_token_ring,     // the node left the token ring, but not group0 yet; we wait until other nodes stop writing to it
```
Does this break compatibility due to the assigned numbers changing?

kbr-scylla replied:
Everything in `topology_state_machine` is behind experimental flag and we're constantly doing incompatible changes in this area.

avikivity replied:
This is bad practice. Sooner or later we'll do it to something that isn't, and that's a very difficult bug to recover from.

kbr-scylla replied:
I remember we did it for CDC, then group 0/schema-over-raft, now we're doing it for topology-over-raft. The code would quickly become completely unmaintainable if we had to introduce a cluster feature or something for every tiny modification we do. The point of experimental flag is to be able to develop a feature quickly, including breaking changes, before we consider it stable for release, no?

gleb-cloudius replied:
> This is bad practice. Sooner or later we'll do it to something that isn't, and that's a very difficult bug to recover from.

The experimental mode exists specifically to not acquire technical dept we will not be able to get rid of afterwords. Some code is introduced with a knowledge that  incompatible modification will be done later.

avikivity replied:
This isn't about a cluster feature, it's about not gratuitously changing an enum's meaning.

gleb-cloudius replied:
The code is at a stage where we not only change enum's meaning, but also removing some values entirely. Even adding a new value without changing existing once requires cluster feature without experimental.

avikivity commented:
Looks good but left some comments.

kbr-scylla commented on message/messaging_service.cc:
```diff
@@ -284,6 +284,26 @@ bool messaging_service::is_same_rack(inet_address addr) const {
+            auto& conn_ref = it->second;
+            conn_ref.server.abort_connection(conn_ref.conn_id);
+        }
+        ms._host_connections.erase(start, end);
```
> What does ban_host() guarantee?

Only what the comment says:
```
// Drops all connections from the given host and prevents further communication from it to happen.
```

Perhaps I should add to the comment that any currently running handlers for this host will not be interrupted so no ambiguity is left. Will follow-up.

We cannot guarantee anything for work that was started concurrently, we can only make guarantees for future events (i.e. no further handlers will be called for RPCs coming from that node).

avikivity replied:
I really meant, what guarantees should ban_host() make.

If calls can continue executing, then a left node can still make changes to the cluster, no?

kbr-scylla replied:
When you have two concurrent operations, you cannot guarantee anything about their order of execution. They can happen one after the other or they can interleave. This in particular applies when we consider the operation of banning a node happening concurrently with some RPC coming from that node. It's just theoretically impossible to give any guarantees here. By the time you call ban_host() the RPC handler might have already finished or it might be at the very end of its execution.

The left node can make some changes to the cluster but that's OK because the decision to make this change was made before it was left (banned). The important thing is that any decision made by it after it left the cluster won't affect the cluster.

avikivity replied:
If ban_host() returns after all handlers have completed, it can guarantee that once it's complete, no operations from that host will take place.

kbr-scylla replied:
We'd have to introduce a bunch of complexity to be able to wait for all handlers that come from a particular hosts to finish (a gate per host or something). But I don't think we need this for anything. If a node made a call when it was still a member it's fine to finish the call. We want to prevent calls that it makes after it was removed, because then it could be using outdated information etc. 

If we find a use case for this we can always extend the mechanism with those gates in a follow-up, but for now I don't see any value in it.

gleb-cloudius replied:
> If ban_host() returns after all handlers have completed, it can guarantee that once it's complete, no operations from that host will take place.

I had the same argument on the patch that introduced `abort_connection()` in the first place https://github.com/scylladb/seastar/pull/1663. Kamil convinced me that for this use case it is fine. You merged the patch, so I assume it was OK with you as well. Later @xemul introduced `server::shutdown()` https://github.com/scylladb/seastar/pull/1671 with the same semantics where I had the same argument and eventually was convinced that it is fine for the intended use case. You merged the patch.

Just the background...

kbr-scylla replied:
@avikivity are the answers satisfactionary? If yes - could you merge?

avikivity replied:
> We'd have to introduce a bunch of complexity to be able to wait for all handlers that come from a particular hosts to finish (a gate per host or something). But I don't think we need this for anything. If a node made a call when it was still a member it's fine to finish the call. We want to prevent calls that it makes after it was removed, because then it could be using outdated information etc.
> 
> If we find a use case for this we can always extend the mechanism with those gates in a follow-up, but for now I don't see any value in it.

So really, we don't need to abort the connections, just prevent new ones.

I don't propose changing anything, just trying to understand. It seems to me that abort without waiting is a halfway measure and either banning new connections, or aborting+waiting would be more self-consistent solutions.

kbr-scylla replied:
> So really, we don't need to abort the connections, just prevent new ones.

If we don't abort the connections, then we would happily keep accepting new RPCs on these existing connections. This is because the "is host ID banned" check appears only once per connection - at the beginning when it's established and we handle the CLIENT_ID verb.

kbr-scylla replied:
@avikivity any chance for getting this merged?

kbr-scylla replied:
@avikivity I answered your last question. Please respond if you have further doubts or need further explanations.

kbr-scylla commented:
v5:
- more details in `ban_host` comment
- the lambda passed to `invoke_on_all` in `ban_host` is no longer a coroutine

Diff: https://github.com/scylladb/scylladb/compare/bddb3af22d69c7a11d8caa67793bab59e1f526d3..70bd6368622d4f9f5dc8ff0df82966a75b7cf05b

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1780/

kbr-scylla commented:
@avikivity could you merge please?

kbr-scylla commented:
@avikivity ping

avikivity commented:
Please rebase, non-trivial conflicts.

kbr-scylla commented:
Hm, probably with fencing PR which was queued 20 minutes ago.

kbr-scylla commented:
I'll rebase once https://github.com/scylladb/scylladb/pull/14285 is merged, which will simplify the "messaging_service: store the node's host ID" commit (we'll be able to make the host_id field non-optional).

avikivity commented:
You're risking a maintainer missing the pings again! But I appreciate doing the right thing.

kbr-scylla commented:
v6:
- `messaging_service` now takes `host_id` in its constructor instead of requiring a separate `set_host_id` call; the host_id inside is non-optional, I moved it into `messaging_service::config` struct.
- the above required a minor movement inside main.cc, to start `messaging_service` after `system_keyspace`. The first commit does that
- rebase
- use `global_token_metadata_barrier` (introduced in fencing PR that was merged in the meantime) instead of regular barrier in `left_token_ring`
- split the addition of `get_cql()` helper to separate commit in manager_client.py

Range-diff:
```range-diff
 1:  b2d3da5622 <  -:  ---------- messaging_service: store the node's host ID
 -:  ---------- >  1:  7f3ad6bd25 main: move messaging_service init after system_keyspace init
 -:  ---------- >  2:  a78cc17bd4 messaging_service: don't use parameter defaults in constructor
 -:  ---------- >  3:  87f65d01b8 messaging_service: store the node's host ID
 2:  5b50d145d3 !  4:  95c726a8df messaging_service: exchange host IDs and map them to connections
    @@ message/messaging_service.cc: shared_ptr<messaging_service::rpc_protocol_client_
              find_and_remove_client(_clients[idx], id, [] (const auto&) { return true; });
          }
      
    -+    auto my_host_id = get_my_host_id();
    ++    auto my_host_id = _cfg.id;
          auto broadcast_address = utils::fb_utilities::get_broadcast_address();
          bool listen_to_bc = _cfg.listen_on_broadcast_address && _cfg.ip != broadcast_address;
          auto laddr = socket_address(listen_to_bc ? broadcast_address : _cfg.ip, 0);
    @@ message/messaging_service.cc: shared_ptr<messaging_service::rpc_protocol_client_
     
      ## message/messaging_service.hh ##
     @@ message/messaging_service.hh: class messaging_service : public seastar::async_sharded_service<messaging_servic
    +     std::vector<scheduling_info_for_connection_index> _scheduling_info_for_connection_index;
          std::vector<tenant_connection_index> _connection_index_for_tenant;
    -     std::optional<locator::host_id> _my_host_id;
      
     +    struct connection_ref;
     +    std::unordered_multimap<locator::host_id, connection_ref> _host_connections;
     +
    +     future<> shutdown_tls_server();
    +     future<> shutdown_nontls_server();
          future<> stop_tls_server();
    -     future<> stop_nontls_server();
    -     future<> stop_client();
 3:  fccd910be4 !  5:  8cf47d76a4 messaging_service: implement host banning
    @@ message/messaging_service.hh: class messaging_service : public seastar::async_sh
          std::unordered_multimap<locator::host_id, connection_ref> _host_connections;
     +    std::unordered_set<locator::host_id> _banned_hosts;
      
    -     future<> stop_tls_server();
    -     future<> stop_nontls_server();
    +     future<> shutdown_tls_server();
    +     future<> shutdown_nontls_server();
     @@ message/messaging_service.hh: class messaging_service : public seastar::async_sharded_service<messaging_servic
          future<table_schema_version> send_schema_check(msg_addr, abort_source&);
      
 4:  7fd7e9ff10 <  -:  ---------- raft topology: `raft_topology_cmd` implicit constructor
 -:  ---------- >  6:  c94c07804d raft topology: `raft_topology_cmd` implicit constructor
 5:  fb9a6c94fb !  7:  b8ddfd9ef9 raft topology: introduce `left_token_ring` state
    @@ service/storage_service.cc: future<> storage_service::topology_state_load(cdc::g
                      on_fatal_internal_error(slogger, ::format("Unexpected state {} for node {}", rs.state, id));
                  }
     @@ service/storage_service.cc: class topology_coordinator {
    -                     builder.del_transition_state()
    +                            .set_version(_topo_sm._topology.version + 1)
                                 .with_node(node.id)
                                 .del("tokens")
     -                           .set("node_state", node_state::left);
    @@ service/storage_service.cc: class topology_coordinator {
                      break;
     +            case node_state::left_token_ring: {
     +                // Wait until other nodes observe the new token ring and stop sending writes to this node.
    -+                // FIXME: change `barrier` to a new command which will wait until existing writes are drained.
    -+                bool exec_command_res;
    -+                std::tie(node, exec_command_res) = co_await exec_global_command(
    -+                        std::move(node), raft_topology_cmd::command::barrier, false);
    -+                if (!exec_command_res) {
    -+                    break;
    ++                {
    ++                    auto id = node.id;
    ++                    auto f = co_await coroutine::as_future(global_token_metadata_barrier(std::move(node)));
    ++                    if (f.failed()) {
    ++                        slogger.error("raft topology: node_state::left_token_ring (node: {}), "
    ++                                      "global_token_metadata_barrier failed, error {}",
    ++                                      id, f.get_exception());
    ++                        break;
    ++                    }
    ++                    node = std::move(f).get();
     +                }
     +
     +                topology_mutation_builder builder(node.guard.write_timestamp());
    @@ service/storage_service.cc: class topology_coordinator {
                  case node_state::decommissioning:
                  case node_state::removing:
     @@ service/storage_service.cc: future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
    -                         result.status = raft_topology_cmd_result::command_status::success;
    -                     }
    -                     break;
    -+                    case node_state::left_token_ring:
    -                     case node_state::left:
    -                     case node_state::none:
    -                     case node_state::removing:
    +                     result.status = raft_topology_cmd_result::command_status::success;
    +                 }
    +                 break;
    ++                case node_state::left_token_ring:
    +                 case node_state::left:
    +                 case node_state::none:
    +                 case node_state::removing:
     
      ## service/topology_state_machine.cc ##
     @@ service/topology_state_machine.cc: static std::unordered_map<node_state, sstring> node_state_to_name_map = {
 6:  201f66a09c !  8:  977680773b raft topology: prepare decommission path for node banning
    @@ service/storage_service.cc: class topology_coordinator {
     +                }
     +
                      // Wait until other nodes observe the new token ring and stop sending writes to this node.
    -                 // FIXME: change `barrier` to a new command which will wait until existing writes are drained.
    -                 bool exec_command_res;
    +                 {
    +                     auto id = node.id;
     @@ service/storage_service.cc: class topology_coordinator {
    -                     break;
    +                     node = std::move(f).get();
                      }
      
     +                // Tell the node to shut down.
    @@ service/storage_service.cc: future<> storage_service::decommission() {
                          // There's nothing smarter we could do. We should not continue operating in this broken
                          // state (we're not a member of the token ring any more).
     @@ service/storage_service.cc: future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
    -                     //co_await sleep_abortable(_db.local().get_config().read_request_timeout_in_ms() * std::chrono::milliseconds(1), _abort_source);
    -                     result.status = raft_topology_cmd_result::command_status::success;
    +                 result.status = raft_topology_cmd_result::command_status::success;
                      break;
    -+                case raft_topology_cmd::command::shutdown:
    -+                    if (_shutdown_request_promise) {
    -+                        std::exchange(_shutdown_request_promise, std::nullopt)->set_value();
    -+                    } else {
    -+                        slogger.warn("raft topology: got shutdown request while not decommissioning");
    -+                    }
    -+                break;
                  }
    -         } catch (...) {
    -             slogger.error("raft topology: raft_topology_cmd failed with: {}", std::current_exception());
    ++            case raft_topology_cmd::command::shutdown:
    ++                if (_shutdown_request_promise) {
    ++                    std::exchange(_shutdown_request_promise, std::nullopt)->set_value();
    ++                } else {
    ++                    slogger.warn("raft topology: got shutdown request while not decommissioning");
    ++                }
    ++                break;
    +         }
    +     } catch (...) {
    +         slogger.error("raft topology: raft_topology_cmd failed with: {}", std::current_exception());
     
      ## service/storage_service.hh ##
     @@ service/storage_service.hh: class storage_service : public service::migration_listener, public gms::i_endpoi
    +     std::optional<shared_future<>> _decomission_result;
          std::optional<shared_future<>> _rebuild_result;
          std::unordered_map<raft::server_id, std::optional<shared_future<>>> _remove_result;
    - 
     +    // During decommission, the node waits for the coordinator to tell it to shut down.
     +    std::optional<promise<>> _shutdown_request_promise;
    -+
    -     future<raft_topology_cmd_result> raft_topology_cmd_handler(sharded<db::system_distributed_keyspace>& sys_dist_ks, raft::term_t term, const raft_topology_cmd& cmd);
    - 
    -     future<> raft_bootstrap(raft::server&);
    +     struct {
    +         raft::term_t term{0};
    +         uint64_t last_index{0};
     
      ## service/topology_state_machine.cc ##
     @@ service/topology_state_machine.cc: std::ostream& operator<<(std::ostream& os, const raft_topology_cmd::command& cmd
    -         case raft_topology_cmd::command::fence_old_reads:
    -             os << "fence_old_reads";
    +         case raft_topology_cmd::command::fence:
    +             os << "fence";
                  break;
     +        case raft_topology_cmd::command::shutdown:
     +            os << "shutdown";
    @@ service/topology_state_machine.cc: std::ostream& operator<<(std::ostream& os, co
     
      ## service/topology_state_machine.hh ##
     @@ service/topology_state_machine.hh: struct raft_topology_cmd {
    -           barrier,         // request to wait for the latest topology
    -           stream_ranges,   // reqeust to stream data, return when streaming is
    -                            // done
    --          fence_old_reads  // wait for all reads started before to complete
    -+          fence_old_reads, // wait for all reads started before to complete
    -+          shutdown         // a decommissioning node should shut down
    +           barrier_and_drain,    // same + drain requests which use previous versions
    +           stream_ranges,        // reqeust to stream data, return when streaming is
    +                                 // done
    +-          fence                 // erect the fence against requests with stale versions
    ++          fence,                // erect the fence against requests with stale versions
    ++          shutdown,             // a decommissioning node should shut down
            };
            command cmd;
      
 7:  f7e56883bd !  9:  737c1b4ae6 raft topology: skip `left_token_ring` state during `removenode`
    @@ Commit message
     
      ## service/storage_service.cc ##
     @@ service/storage_service.cc: class topology_coordinator {
    -         co_return std::pair{retake_node(std::move(guard), node.id), res};
    +         co_return retake_node(std::move(guard), node.id);
          };
      
     +    future<> remove_from_group0(const raft::server_id& id) {
    @@ service/storage_service.cc: class topology_coordinator {
     +                    auto next_state = node.rs->state == node_state::decommissioning
     +                                        ? node_state::left_token_ring : node_state::left;
                          builder.del_transition_state()
    +                            .set_version(_topo_sm._topology.version + 1)
                                 .with_node(node.id)
                                 .del("tokens")
     -                           .set("node_state", node_state::left_token_ring);
 8:  2f53f68135 = 10:  63229e48e8 raft topology: ban left nodes
 9:  9084a493cd = 11:  e02249f0cd test: pylib: ScyllaCluster: server pause/unpause API
 -:  ---------- > 12:  ae92932240 test: pylib: manager_client: `get_cql()` helper
10:  fc5a4259d8 ! 13:  279a109ce0 test: add node banning test
    @@ Commit message
     
         Check that it is not able to perform queries once it unpauses.
     
    - ## test/pylib/manager_client.py ##
    -@@ test/pylib/manager_client.py: class ManagerClient():
    -             logger.debug("refresh driver node list")
    -             self.ccluster.control_connection.refresh_node_list_and_token_map()
    - 
    -+    def get_cql(self) -> CassandraSession:
    -+        assert self.cql
    -+        return self.cql
    -+
    -     async def before_test(self, test_case_name: str) -> None:
    -         """Before a test starts check if cluster needs cycling and update driver connection"""
    -         logger.debug("before_test for %s", test_case_name)
    -
      ## test/topology_experimental_raft/test_node_isolation.py (new) ##
     @@
     +#
11:  70bd636862 = 14:  b38dcba6ed test: pylib: increase checking period for `get_alive_endpoints`
```

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1946/

kbr-scylla commented:
> CI state ABORTED - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1946/

```
[2023-06-20T12:11:35.012Z] ERROR: DTest failed: https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/2354/
[2023-06-20T12:11:38.994Z] Finished: ABORTED
```

https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/2354/
```
hudson.AbortException: Spot termination
```

rekicking

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1949/

kbr-scylla commented:
@avikivity ready for merging again :)

kbr-scylla commented:
@avikivity ping
